components/assistive-playwright-client/src/screenReaderClient.ts

```diff
@@ -0,0 +1,340 @@
+   * as soon as they arrive on the websocket, and can be removed by
+   * {@link ScreenReaderClient.clearMessages} and
+   * {@link ScreenReaderClient.waitForMessage}.
+   */
+  messages: string[] = [];
+
+  /**
+   * Connects to a websocket url and returns a corresponding
+   * instance of {@link ScreenReaderClient} as soon as it is
+   * connected.
+   * @param url - Websocket URL such as `ws://127.0.0.1:7779/screen-reader`
+   */
+  static async create(url: string): Promise<ScreenReaderClient> {
+    const result = new ScreenReaderClient(url);
+    await new Promise((resolve, reject) => {
+      result._emitter.once("open", resolve);
+      result._emitter.once("error", reject);
+    });
+    return result;
+  }
+
+  private constructor(url: string) {
+    const socket = new Websocket(url);
+    this.socket = socket;
+    socket.onmessage = arg => {
+      const data = arg.data.toString("utf8");
+      this.messages.push(data);
+      this._emitter.emit("message", data);
+    };
+    socket.onopen = () => {
+      this.connected = true;
+      this._emitter.emit("open");
+    };
+    socket.onclose = () => {
+      this.connected = false;
+      this._emitter.emit("close");
+    };
+    socket.onerror = error => {
+      this._emitter.emit("error", error);
+    };
+  }
+
+  /**
+   * Disconnects the underlying websocket, thus preventing
+   * any further message from being received.
+   */
+  disconnect(): void {
+    if (this.connected) {
+      this.socket.close();
+    }
+  }
+
+  /**
+   * Execute the `condition` function on each received message until it returns a
+   * truthy value or the specified timeout occurs.
+   * If there is a timeout, the promise returned by this function is rejected.
+   * @param condition - function that is called for each message as long as it returns a falsy value. When it returns a
+   * truthy value, the process is stopped and that value is returned by
+   * waitForMessageCondition.
+   * @param options - `includeCurrent`: if true (the default), the `condition` function
+   * is first called for already received messages that are in {@link ScreenReaderClient.messages | messages}.
+   * Otherwise, it is only called on new messages, not yet received at the time `waitForMessageCondition` is called.
+   * `timeout`: the time in milliseconds to wait before the promise returned by
+   * `waitForMessageCondition` is rejected (if the condition does not return a
+   * truthy value during that time). Defaults to 10000ms (10s).
+   */
+  waitForMessageCondition<T>(
+    condition: (
+      value: string,
+      index: number,
+      messages: string[]
+    ) => T | undefined,
+    options: {
+      includeCurrent?: boolean;
+      timeout?: number;
+    } = {}
+  ): Promise<T> {
+    const { includeCurrent = true, timeout = 10000 } = options;
+    return new Promise<T>((resolve, reject) => {
+      if (includeCurrent) {
+        for (let msg = this.messages, i = 0, l = msg.length; i < l; i++) {
```

Comment:
It's a matter of taste (and coding preferences), but for this kind of loop (so here and in several places in this PR), I would have preferred  the `for...of` or `for...in` usage which is quite concise:

for... in here as your need the index in the body:
```
const messages = this.messages;
for(const i in messages) {
   const result = condition(messages[i], i, messages);
   ...
}
```

Reply:
I could have used `for ... of` if I did not need the index.
But `for ... in` is for the properties of any object, and there is no guarrantee about the order of iteration: cf https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#array_iteration_and_for...in

Reply:
Right, I didn't remember about the order. So it's better to let it like this.

Just a note: the order of a for...in is in the spec of the ES2020, so the browsers should end to respect it (it's probably already the case, but not guarantee).

Reply:
Note that this code is not supposed to run in a browser, but in node.js, which, indeed, probably respects the order defined by the spec of ES2020, but I prefer having a clear code for which it is not necessary to wonder if the order will be correct.